### PR TITLE
[JENKINS-58093] - Support for LabelAtoms

### PR DESCRIPTION
**Pull request by varyvol**
[JENKINS-58093](https://issues.jenkins-ci.org/browse/JENKINS-58093)

LabelAtoms should work with CasC. I'm trying to implement it but I'm kind of stuck since I'm not able to properly export the labels configuration. In theory the DescribableListAttribute should work, but I'm not able to do it and exporting generates the following error:
`FAILED TO EXPORT\n      hudson.model.labels.LabelAtom#properties: No configurator found for type class hudson.model.labels.LabelAtomProperty`

Can someone please help me with moving this forward?

### Your checklist for this pull request

🚨 Please review the [guidelines for contributing](../blob/master/docs/CONTRIBUTING.md) to this repository.

- [X] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side) and not your master branch!
- [X] Ensure that the pull request title represents the desired changelog entry
- [X] Please describe what you did
- [X] Link to relevant issues in GitHub or in [Jenkins JIRA](https://issues.jenkins-ci.org)
- [X] Link to relevant pull requests, esp. upstream and downstream changes
- [X] Did you provide a test-case? That demonstrates feature works or fixes the issue.

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
No review ETA right now, underwater with the GSoC prepwork

**Comment by jetersen** [comment]
CI is failing with failed export of `LabelAtomProperty`

**Comment by varyvol** [comment]
@jetersen yes, but I don't understand why. I mention it in the description, I'd like to get some help if possible.

File: plugin/src/main/java/io/jenkins/plugins/casc/core/LabelAtomConfigurator.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    @NonNull
+    @Override
+    public Set<Attribute<LabelAtom, ?>> describe() {
+        return new HashSet<>(Arrays.asList(
+            new Attribute<LabelAtom, String>("name", String.class)
+                .getter(Label::getName)
+                .setter( noop() ),
+            new DescribableListAttribute<LabelAtom, LabelAtomProperty>("properties", LabelAtomProperty.class)
```
Comment (jetersen) [inline plugin/src/main/java/io/jenkins/plugins/casc/core/LabelAtomConfigurator.java:43]: `LabelAtomProperty` is not describable so you need a configurator for it too.

Comment (jetersen) [reply plugin/src/main/java/io/jenkins/plugins/casc/core/LabelAtomConfigurator.java:43]: The class is abstract with no descriptior definition.
https://github.com/jenkinsci/jenkins/blob/6baeb65e2c2cec8e2a22b4fe4e71c803c7373e85/core/src/main/java/hudson/model/labels/LabelAtomPropertyDescriptor.java#L43

That is sure to cause trouble.

Comment (varyvol) [reply plugin/src/main/java/io/jenkins/plugins/casc/core/LabelAtomConfigurator.java:43]: Isn't it if it extends AbstractDescribableImpl?

Comment (varyvol) [reply plugin/src/main/java/io/jenkins/plugins/casc/core/LabelAtomConfigurator.java:43]: It is designed that way for plugins to contribute with label properties.

**Comment by timja** [comment]
why close? did you figure out how to change core for this?

**Comment by varyvol** [comment]
@timja no, I have not been able to make it work, that's why I closed.

**Comment by varyvol** [comment]
@jetersen @timja I've added the configurator for LabelAtomProperty, but it's now complaining that the class has no DataboundConstructor, which is true.

But I would expect the actual class implementing the LabelAtomProperty to be used? Or is there anything to change in the configurator, in the lookUp code or in how those classes are implemented in the core for it to work?

Any suggestions, ideas, proposals more than welcome.

File: plugin/src/main/java/io/jenkins/plugins/casc/core/LabelAtomConfigurator.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    @NonNull
+    @Override
+    public Set<Attribute<LabelAtom, ?>> describe() {
+        return new HashSet<>(Arrays.asList(
+            new Attribute<LabelAtom, String>("name", String.class)
+                .getter(Label::getName)
+                .setter( noop() ),
+            new MultivaluedAttribute<LabelAtom, LabelAtomProperty>("properties", LabelAtomProperty.class)
```
Comment (timja) [inline plugin/src/main/java/io/jenkins/plugins/casc/core/LabelAtomConfigurator.java:44]: you can create your object that you can use instead if that makes sense?
https://github.com/jenkinsci/configuration-as-code-plugin/blob/master/plugin/src/main/java/io/jenkins/plugins/casc/core/HudsonPrivateSecurityRealmConfigurator.java#L45

i.e. UserWithPassword,

I haven't looked closely into this though

Comment (varyvol) [reply plugin/src/main/java/io/jenkins/plugins/casc/core/LabelAtomConfigurator.java:44]: But how can I create my object if I don't know what each implementation provides? One LabelAtomProperty might have a couple of String fields, another one might have a String and an integer for example.

Comment (timja) [reply plugin/src/main/java/io/jenkins/plugins/casc/core/LabelAtomConfigurator.java:44]: Does it work if the implementation LabelAtomProperty has a databoundcontructor on it?

i.e. maybe it just needs adding in a few places

Comment (varyvol) [reply plugin/src/main/java/io/jenkins/plugins/casc/core/LabelAtomConfigurator.java:44]: No, you can see how "TestProperty" has a DataboundConstructor and it's not working.

**Comment by varyvol** [comment]
@jetersen @timja @oleg-nenashev it should be ready now.

**Comment by varyvol** [comment]
@jetersen thanks for such a quick merge! Do you think having this released is a possibility?

**Comment by jetersen** [comment]
I'll get a release rolling in the evening if nobody gets to it before me, 😉 @timja

**Comment by varyvol** [comment]
@jetersen awesome, thanks!

**Comment by timja** [comment]
hah, i can do it now

File: test-harness/src/test/resources/io/jenkins/plugins/casc/core/ConfigureNodes.yml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+jenkins:
```
Comment (jglick) [inline test-harness/src/test/resources/io/jenkins/plugins/casc/core/ConfigureNodes.yml:1]: This file seems unused by the new test?

Comment (jglick) [reply test-harness/src/test/resources/io/jenkins/plugins/casc/core/ConfigureNodes.yml:1]: also see #2130

**Comment by mawinter69** [comment]
Seems the description of the atoms is not saved and thus can't be configured via CasC

**Pull request merged**